# Movies!

**Because who doesn't love a good movie?**

We've all been there: it's another Friday night, and the COVID pandemic has limited our options for evening entertainment. Go to a party at a friend's house? Can't -- that'd be irresponsible and potentially dangerous. Check out a night club? Nope (let's be real, that was an awful option even before the pandemic hit).

And so, we're resorting once again to -- movie night! Make some stove top popcorn and plop down for 90 - 180 minutes of unbridled cinematic joy in the comfort of your own home (and pajamas).

But how can we possibly make the best choice possible when there are so many options? Nothing is worse than watching the ending credits flash on the screen and thinking (or exclaiming aloud, in extreme instances), "MAN, that sucked!" Here to help is a **movie recommender engine** using data obtained from <https://github.com/rashida048/Datasets/blob/master/movie_dataset.csv>

## Part I: Exploration

Let's read in the `dataset` and do some preliminary analysis.

In [1]:
# Make necessary imports
import pandas as pd
import numpy as np

In [2]:
# Read in the CSV file and check the head
movies = pd.read_csv('../datasets/movie_dataset.csv')
movies.head()

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

In [3]:
# How many rows and columns does our dataset have?
movies.shape

(4803, 24)

In [4]:
# What are the names of the columns in our dataset?
movies.columns

Index(['index', 'budget', 'genres', 'homepage', 'id', 'keywords',
       'original_language', 'original_title', 'overview', 'popularity',
       'production_companies', 'production_countries', 'release_date',
       'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title',
       'vote_average', 'vote_count', 'cast', 'crew', 'director'],
      dtype='object')

In [5]:
# How many nulls are in our dataset?
movies.isnull().sum().sort_values(ascending = False)

homepage                3091
tagline                  844
keywords                 412
cast                      43
director                  30
genres                    28
overview                   3
runtime                    2
release_date               1
popularity                 0
budget                     0
id                         0
original_language          0
original_title             0
production_countries       0
production_companies       0
crew                       0
revenue                    0
spoken_languages           0
status                     0
title                      0
vote_average               0
vote_count                 0
index                      0
dtype: int64

In [6]:
# What datatype is each column?
movies.dtypes

index                     int64
budget                    int64
genres                   object
homepage                 object
id                        int64
keywords                 object
original_language        object
original_title           object
overview                 object
popularity              float64
production_companies     object
production_countries     object
release_date             object
revenue                   int64
runtime                 float64
spoken_languages         object
status                   object
tagline                  object
title                    object
vote_average            float64
vote_count                int64
cast                     object
crew                     object
director                 object
dtype: object

**Let's look at each column one by one to see what the data looks like**

# Part II: Step by Step Examination and Cleaning

### Index

In [7]:
# Preview column index
movies[['index']]

index
0         0
1         1
2         2
3         3
4         4
...     ...
4798   4798
4799   4799
4800   4800
4801   4801
4802   4802

[4803 rows x 1 columns]

This data simply matches the already-existing index data of each row. This column can and should be `dropped`.

In [8]:
# Drop this useless column from the dataframe
movies.drop(columns = 'index', inplace = True)

### **Genres**

In [9]:
# Preview column genres
movies[['genres']]

genres
0     Action Adventure Fantasy Science Fiction
1                     Adventure Fantasy Action
2                       Action Adventure Crime
3                  Action Crime Drama Thriller
4             Action Adventure Science Fiction
...                                        ...
4798                     Action Crime Thriller
4799                            Comedy Romance
4800             Comedy Drama Romance TV Movie
4801                                       NaN
4802                               Documentary

[4803 rows x 1 columns]

It looks like this column contains each genre that can be categorically applied to each movie, with no sperator betweemn the genres (i.e., `,`). Let's look to see what the longest `genre` cell in this dataset look like.

In [10]:
# Sort movies by longest genre
movies.genres.str.len().sort_values(ascending = False)

1617    51.0
3208    50.0
305     50.0
1277    50.0
168     50.0
        ... 
4674     NaN
4681     NaN
4714     NaN
4716     NaN
4801     NaN
Name: genres, Length: 4803, dtype: float64

In [11]:
movies.iloc[1617]['genres']

'Action Adventure Animation Science Fiction Thriller'

This movie looks like the greatest one on the list! Who wouldn't want to watch an `Action/Adventure/Animation/Science Fiction/Thriller?`

When taking a glance at the `NaNs` in this dataset, we can see that there are `28` blank cells in the `genre` category. Let's clear those out now.

In [12]:
# Drop rows in which the 'genre' values are NaNs.
movies = movies[movies['genres'].notna()]

# Check work
movies[movies['genres'].isna()]

Empty DataFrame
Columns: [budget, genres, homepage, id, keywords, original_language, original_title, overview, popularity, production_companies, production_countries, release_date, revenue, runtime, spoken_languages, status, tagline, title, vote_average, vote_count, cast, crew, director]
Index: []

[0 rows x 23 columns]

**How many unique genres are there in the dataset?**

In [13]:
# Create an empty list called 'genres'
genres = []

# Create a list of indices as long as the column 'genres' in the dataframe
for i in range(0, len(movies['genres'])):
    
    # Isolate each cell one by one and create a list of genres in the cell
    cell = movies.iloc[i]['genres'].split(' ')
    
    # Loop through each list to look for unique genres
    for genre in cell:
        
        # Add unique genres to the genre list
        if genre not in genres:
            genres.append(genre)
            
# Take a look at the genres list and how many unique genres are in it
print(f'There are {len(genres)} unique genres in the dataset, and they are:')
print(genres)

There are 22 unique genres in the dataset, and they are:
['Action', 'Adventure', 'Fantasy', 'Science', 'Fiction', 'Crime', 'Drama', 'Thriller', 'Animation', 'Family', 'Western', 'Comedy', 'Romance', 'Horror', 'Mystery', 'History', 'War', 'Music', 'Documentary', 'Foreign', 'TV', 'Movie']


Looking above, it looks like `Science Fiction` has been split up into `Science` and `Fiction`. This might be ignorable if this had happened to `Romantic Comedy`, but we can't ignore the 2nd best movie `genre` out there (the first being `Horror`, of course). 

We can also see that there are 2 genres, one being `TV` and one being `Movie`, that are included in this dataset. This is likely a split of the genre `TV Movie`. If the genre `TV Movie` were shortened to just `TV`, no context would be lost. That said, let's drop `Movie` from final `genres` list.

Let's re-write the function above to fix this. This will be necessary in `binarizing` this column later.

In [14]:
# Create an empty list called 'genres'
genres = []

# Create a list of indices as long as the column 'genres' in the dataframe
for i in range(0, len(movies['genres'])):
    
    # Isolate each cell one by one and create a list of genres in the cell
    cell = movies.iloc[i]['genres'].split(' ')
    
    # Look for 'Science' and 'Fiction' in each cell
    if 'Science' and 'Fiction' in cell:
        
        # Drop 'Science'
        cell.remove('Science')
        
        # Drop 'Fiction'
        cell.remove('Fiction')
        
        #Add 'Science Fiction'
        cell.append('Science_Fiction')
    
    # Loop through each list to look for unique genres
    for genre in cell:
        
        # Skip over the genre 'Movie'
        if genre == 'Movie':
            pass

        # Add unique genres to the genre list
        elif genre not in genres:
            genres.append(genre)
            
# Take a look at the genres list and how many unique genres are in it
print(f'There are {len(genres)} unique genres in the dataset, and they are:')
print(genres)

There are 20 unique genres in the dataset, and they are:
['Action', 'Adventure', 'Fantasy', 'Science_Fiction', 'Crime', 'Drama', 'Thriller', 'Animation', 'Family', 'Western', 'Comedy', 'Romance', 'Horror', 'Mystery', 'History', 'War', 'Music', 'Documentary', 'Foreign', 'TV']


**Now, let's binarize the genres and drop the genres column from the dataset**

In [15]:
def binarize_genres(df):
    
    # Create a new dataframe with a column for each unique genre in the list above
    df2 = pd.DataFrame(columns = [genre for genre in genres], index = df.index, data = 0)
        
    # Create a list of indices as long as the column 'genres' in the dataframe
    for i in range(0, len(df['genres'])):
    
        # Isolate each cell one by one and create a list of genres in the cell
        cell = df.iloc[i]['genres'].split(' ')
    
        # Look for 'Science' and 'Fiction' in each cell
        if 'Science' and 'Fiction' in cell:
        
            # Drop 'Science'
            cell.remove('Science')
        
            # Drop 'Fiction'
            cell.remove('Fiction')
        
            #Add 'Science Fiction'
            cell.append('Science_Fiction')
    
        # Loop through the cell for each genre
        for genre in cell:
        
            # Loop through each column of df2
            for column in df2.columns:
                
                # Skip over the genre 'Movie'
                # This step is redundant, as we've already dropped 'Movies' from the 'genres' list, which
                # /n was used to create all the columns in df2. This has been included for continuity's sake.
                if genre == 'Movie':
                    pass
                
                # Look for matching genres
                elif genre == column:
                    
                    # Mark the genre as a 1
                    df2[column].iloc[i] += 1
    
    # Create a list variable called frames for our 2 dataframes
    frames = [df, df2]
    
    # Concatenate the two dataframes
    final_df = pd.concat(frames, axis = 1, sort = False)
    
    return final_df

In [16]:
# Call function and change 
movies = binarize_genres(movies)

# Drop the genres column now that we've binarized its contents
movies.drop(columns = 'genres', inplace = True)

### **Home page**

In [17]:
movies[['homepage']]

homepage
0                           http://www.avatarmovie.com/
1          http://disney.go.com/disneypictures/pirates/
2           http://www.sonypictures.com/movies/spectre/
3                    http://www.thedarkknightrises.com/
4                  http://movies.disney.com/john-carter
...                                                 ...
4797                                                NaN
4798                                                NaN
4799                                                NaN
4800  http://www.hallmarkchannel.com/signedsealeddel...
4802                                                NaN

[4775 rows x 1 columns]

This column isn't important for a recommender engine and can be dropped.

In [18]:
# Drop column 'homepage' from dataset
movies.drop(columns = 'homepage', inplace = True)

### **ID**

In [19]:
movies[['id']]

id
0      19995
1        285
2     206647
3      49026
4      49529
...      ...
4797   67238
4798    9367
4799   72766
4800  231617
4802   25975

[4775 rows x 1 columns]

This column isn't important for a recommender engine and can be dropped.

In [20]:
# Drop column 'id' from dataset
movies.drop(columns = 'id', inplace = True)

### **Keywords**

In [21]:
movies[['keywords']]

keywords
0     culture clash future space war space colony so...
1     ocean drug abuse exotic island east india trad...
2            spy based on novel secret agent sequel mi6
3     dc comics crime fighter terrorist secret ident...
4     based on novel mars medallion space travel pri...
...                                                 ...
4797                                                NaN
4798  united states\u2013mexico barrier legs arms pa...
4799                                                NaN
4800  date love at first sight narration investigati...
4802               obsession camcorder crush dream girl

[4775 rows x 1 columns]

This looks like a tricky column. Not only are there 412 `NaNs` in this column, but the keywords here look pretty subjective. Who decided on these keywords? The Academy at large? Probably not. Due to its subjectivity and missing data, this column looks droppable from the final dataset.

In [22]:
# Drop column 'keywords' from dataset
movies.drop(columns = 'keywords', inplace = True)

### **Original Language**

In [23]:
movies[['original_language']]

original_language
0                   en
1                   en
2                   en
3                   en
4                   en
...                ...
4797                en
4798                es
4799                en
4800                en
4802                en

[4775 rows x 1 columns]

In [24]:
# Check unique values of this column
movies.original_language.unique()

array(['en', 'ja', 'fr', 'zh', 'es', 'de', 'hi', 'ru', 'ko', 'te', 'cn',
       'it', 'nl', 'ta', 'sv', 'th', 'da', 'xx', 'hu', 'cs', 'pt', 'is',
       'tr', 'nb', 'af', 'pl', 'he', 'ar', 'vi', 'ky', 'id', 'ro', 'fa',
       'no', 'sl', 'ps', 'el'], dtype=object)

It's unclear what a majority of these abbreviations stand for. There's no key for what these might stand for on the website from which this dataset was pulled, which isn't helpful. Should  we drop this column from the dataset? Or attempt to find a key for this online somewhere? Probably not -- a different key might have been created by someone who would abbreviate languages differently, which could corrupt the accuracy of this dataset. 

In [25]:
movies.original_language.describe()

count     4775
unique      37
top         en
freq      4477
Name: original_language, dtype: object

It looks like there are 298 movies of the 4,803 in this dataset that aren't in English. In the interest of time, simplicity, and accuracy, it may be worth changing this project to a movie recommendation enginge in which all movies are in English.

In [26]:
# Look at the dataset for all non-English movies
movies[movies['original_language'] != 'en']

budget original_language                      original_title  \
97    15000000                ja                              シン・ゴジラ   
235   97250400                fr         Astérix aux Jeux Olympiques   
317   94000000                zh                               金陵十三釵   
474          0                fr                           Évolution   
492    8000000                es  Don Gato: El inicio de la pandilla   
...        ...               ...                                 ...   
4739         0                fr         I Love You, Don't Touch Me!   
4751         0                pt            Gabriela, Cravo e Canela   
4790         0                fa                               دایره   
4792     20000                ja                                 キュア   
4798    220000                es                         El Mariachi   

                                               overview  popularity  \
97    From the mind behind Evangelion comes a hit la...    9.476999   
235   Astérix and Obélix have to win the Olympic Gam...   20.344364   
317   A Westerner finds refuge with a group of women...   12.516546   
474   11-year-old Nicolas lives with his mother in a...    3.300061   
492   Top Cat has arrived to charm his way into your...    0.719996   
...                                                 ...         ...   
4739  The story of a 25 year old virgin girl, lookin...    0.020839   
4751  In 1925, Gabriela becomes cook, mistress, and ...    0.557602   
4790  Various women struggle to function in the oppr...    1.193779   
4792  A wave of gruesome murders is sweeping Tokyo. ...    0.212443   
4798  El Mariachi just wants to play his guitar and ...   14.269792   

                                   production_companies  \
97    [{"name": "Cine Bazar", "id": 5896}, {"name": ...   
235   [{"name": "Constantin Film", "id": 47}, {"name...   
317   [{"name": "Beijing New Picture Film Co. Ltd.",...   
474   [{"name": "Ex Nihilo", "id": 3307}, {"name": "...   
492   [{"name": "Anima Estudios", "id": 9965}, {"nam...   
...                                                 ...   
4739                                                 []   
4751  [{"name": "United Artists", "id": 60}, {"name"...   
4790  [{"name": "Jafar Panahi Film Productions", "id...   
4792             [{"name": "Daiei Studios", "id": 881}]   
4798           [{"name": "Columbia Pictures", "id": 5}]   

                                   production_countries release_date  \
97              [{"iso_3166_1": "JP", "name": "Japan"}]   2016-07-29   
235   [{"iso_3166_1": "BE", "name": "Belgium"}, {"is...   2008-01-13   
317   [{"iso_3166_1": "CN", "name": "China"}, {"iso_...   2011-12-15   
474   [{"iso_3166_1": "BE", "name": "Belgium"}, {"is...   2015-09-14   
492   [{"iso_3166_1": "IN", "name": "India"}, {"iso_...   2015-10-30   
...                                                 ...          ...   
4739  [{"iso_3166_1": "US", "name": "United States o...   1997-01-21   
4751           [{"iso_3166_1": "BR", "name": "Brazil"}]   1983-03-24   
4790             [{"iso_3166_1": "IR", "name": "Iran"}]   2000-09-08   
4792            [{"iso_3166_1": "JP", "name": "Japan"}]   1997-11-06   
4798  [{"iso_3166_1": "MX", "name": "Mexico"}, {"iso...   1992-09-04   

        revenue  runtime  ... Comedy Romance Horror Mystery  History  War  \
97     77000000    120.0  ...      0       0      1       0        0    0   
235   132900000    116.0  ...      1       0      0       0        0    0   
317    95311434    145.0  ...      0       0      0       0        1    1   
474           0     81.0  ...      0       0      1       1        0    0   
492           0     89.0  ...      1       0      0       0        0    0   
...         ...      ...  ...    ...     ...    ...     ...      ...  ...   
4739          0     86.0  ...      1       1      0       0        0    0   
4751          0     99.0  ...      0       1      0       0        0    0   
4790          0     90.

After exploring this list on non-English movies further, my decision to drop these from the dataset is validated. Although the `overview` column for these movies are in English, what good is a movie recommendation for a film called "دایره" for someone who doesn't speak that native language? I searched for "دایره" on Google to see if I was able to find the movie described in this dataset, but my search turned up no results. 

Let's drop these movies from our dataset.

In [27]:
# Drop all non-English movies
movies = movies[movies['original_language'] == 'en']

In [28]:
# Now that all movies are in English, we can drop this column from our dataframe
movies.drop(columns = 'original_language', inplace = True)

### **Original Title**

In [29]:
movies[['original_title']]

original_title
0                                       Avatar
1     Pirates of the Caribbean: At World's End
2                                      Spectre
3                        The Dark Knight Rises
4                                  John Carter
...                                        ...
4796                                    Primer
4797                                    Cavite
4799                                 Newlyweds
4800                 Signed, Sealed, Delivered
4802                         My Date with Drew

[4477 rows x 1 columns]

Nothing to see here. Using any `Natural Language Processing` or `Binarizing` here would be a mistake, unless people prefer to make watching decisions based on a movie's title and title alone.

What we can do, however, is set this column to be the dataframe's `index` column.

In [30]:
#Set index
movies.set_index('original_title', inplace = True)

# Check work
movies.head()

budget  \
original_title                                        
Avatar                                    237000000   
Pirates of the Caribbean: At World's End  300000000   
Spectre                                   245000000   
The Dark Knight Rises                     250000000   
John Carter                               260000000   

                                                                                   overview  \
original_title                                                                                
Avatar                                    In the 22nd century, a paraplegic Marine is di...   
Pirates of the Caribbean: At World's End  Captain Barbossa, long believed to be dead, ha...   
Spectre                                   A cryptic message from Bond’s past sends him o...   
The Dark Knight Rises                     Following the death of District Attorney Harve...   
John Carter                               John Carter is a war-weary, former military ca...   

                                          popularity  \
original_title                                         
Avatar                                    150.437577   
Pirates of the Caribbean: At World's End  139.082615   
Spectre                                   107.376788   
The Dark Knight Rises                     112.312950   
John Carter                                43.926995   

                                                                       production_companies  \
original_title                                                                                
Avatar                                    [{"name": "Ingenious Film Partners", "id": 289...   
Pirates of the Caribbean: At World's End  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
Spectre                                   [{"name": "Columbia Pictures", "id": 5}, {"nam...   
The Dark Knight Rises                     [{"name": "Legendary Pictures", "id": 923}, {"...   
John Carter                                     [{"name": "Walt Disney Pictures", "id": 2}]   

                                                                       production_countries  \
original_title                                                                                
Avatar                                    [{"iso_3166_1": "US", "name": "United States o...   
Pirates of the Caribbean: At World's End  [{"iso_3166_1": "US", "name": "United States o...   
Spectre                                   [{"iso_3166_1": "GB", "name": "United Kingdom"...   
The Dark Knight Rises                     [{"iso_3166_1": "US", "name": "United States o...   
John Carter                               [{"iso_3166_1": "US", "name": "United States o...   

                                         release_date     revenue  runtime  \
original_title                                                               
Avatar                                     2009-12-10  2787965087    162.0   
Pirates of the Caribbean: At World's End   2007-05-19   961000000    169.0   
Spectre                                    2015-10-26   880674609    148.0   
The Dark Knight Rises                      2012-07-16  1084939099    165.0   
John Carter                                2012-03-07   284139100    132.0   

                                                                           spoken_languages  \
original_title                                                                                
Avatar                                    [{"iso_639_1": "en", "name": "English"}, {"iso...   
Pirates of the Caribbean: At World's End           [{"iso_639_1": "en", "name": "English"}]   
Spectre                                   [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...   
The Dark Knight Rises                              [{"iso_639_1": "en", "name": "English"}]   
John Carter                                        [{"iso_639_1": "en", "name": "English"}]   

                                            status  ...

### **Overview**

In [31]:
movies[['overview']]

overview
original_title                                                                             
Avatar                                    In the 22nd century, a paraplegic Marine is di...
Pirates of the Caribbean: At World's End  Captain Barbossa, long believed to be dead, ha...
Spectre                                   A cryptic message from Bond’s past sends him o...
The Dark Knight Rises                     Following the death of District Attorney Harve...
John Carter                               John Carter is a war-weary, former military ca...
...                                                                                     ...
Primer                                    Friends/fledgling entrepreneurs invent a devic...
Cavite                                    Adam, a security guard, travels from Californi...
Newlyweds                                 A newlywed couple's honeymoon is upended by th...
Signed, Sealed, Delivered                 "Signed, Sealed, Delivered" introduces a dedic...
My Date with Drew                         Ever since the second grade when he first saw ...

[4477 rows x 1 columns]

This column looks more insightful than the `keywords` column. We can use `Natural Language Processing` when creating the model for this dataset.

Before doing so, let's check one more time for `NaNs` in this column. There were `3` of them for `overview` initially, but we've dropped some data from the dataframe `movies` at this point in this jupyter notebook.

In [32]:
# Check once more for nulls in the 'overview' column
movies['overview'].isna().sum()

1

In [33]:
# Let's check to see what this one null overview is from
movies[movies['overview'].isna()]

budget overview  popularity  \
original_title                                             
To Be Frank, Sinatra at 100       2      NaN    0.050625   

                                                         production_companies  \
original_title                                                                  
To Be Frank, Sinatra at 100  [{"name": "Eyeline Entertainment", "id": 60343}]   

                                                         production_countries  \
original_title                                                                  
To Be Frank, Sinatra at 100  [{"iso_3166_1": "GB", "name": "United Kingdom"}]   

                            release_date  revenue  runtime spoken_languages  \
original_title                                                                
To Be Frank, Sinatra at 100   2015-12-12        0      NaN               []   

                               status  ... Comedy Romance  Horror  Mystery  \
original_title                         ...                                   
To Be Frank, Sinatra at 100  Released  ...      0       0       0        0   

                            History War Music  Documentary  Foreign  TV  
original_title                                                           
To Be Frank, Sinatra at 100       0   0     0            1        0   0  

[1 rows x 37 columns]

Ahhh yes, the critically acclaimed "To Be Frank, Sinatra at 100", the rave documentary with the $2 budget. 

On a serious note -- if we remove this film from our dataframe, we'll be able to use NLP for 100% of the `overviews`.

In [34]:
# Remove the movie 'To Be Frank, Sinartra at 100' from the dataframe
movies.drop(labels = 'To Be Frank, Sinatra at 100', inplace = True)

In [35]:
# Import Tokenizer
from nltk.tokenize import RegexpTokenizer

# Import Lemmatizer
from nltk.stem import WordNetLemmatizer

# Import stopwords.
from nltk.corpus import stopwords

In [36]:
# Write a function to tokenize and then lemmatize the overview
def lemmatize_overview(df):
    
    # Create an empty list that will ultimately be a new column of lemmas
    overview_lemmas = []

    # Instantiate a tokenizer
    # Include the 'r\w+' to ignore spaces and punctuation
    tokenizer = RegexpTokenizer(r'\w+')
    
    # Instantiate a lemmatizer
    lemmatizer = WordNetLemmatizer()
    
    for cell in df['overview']:
        
        # Tokenize the individual overview
        overview_tokens = tokenizer.tokenize(cell.lower())
        
        # Lemmatize the individual overview
        overview_lemma = [lemmatizer.lemmatize(token) for token in overview_tokens]
        
        # Append the lemmatized cell to the empty list titles 'overview_lemmas'
        overview_lemmas.append(overview_lemma)
    
    # Drop the original overview column from the dataframe
    df.drop(columns = 'overview', inplace = True)
    
    # Replace column with the lemmas, rename it 'overview'
    df['overview'] = overview_lemmas
    
    # Return df
    return df


In [37]:
# Call the function and see if it worked
movies = lemmatize_overview(movies)
movies['overview']

original_title
Avatar                                      [in, the, 22nd, century, a, paraplegic, marine...
Pirates of the Caribbean: At World's End    [captain, barbossa, long, believed, to, be, de...
Spectre                                     [a, cryptic, message, from, bond, s, past, sen...
The Dark Knight Rises                       [following, the, death, of, district, attorney...
John Carter                                 [john, carter, is, a, war, weary, former, mili...
                                                                  ...                        
Primer                                      [friend, fledgling, entrepreneur, invent, a, d...
Cavite                                      [adam, a, security, guard, travel, from, calif...
Newlyweds                                   [a, newlywed, couple, s, honeymoon, is, upende...
Signed, Sealed, Delivered                   [signed, sealed, delivered, introduces, a, ded...
My Date with Drew                           [

Now that we've `tokenized` and `lemmatized` the `overview` column, we can use one of the vectorizer alogrorithms (either `CountVectorizer` or `TFIDFVectorizer` to process the language for modeling. We'll do that in the next notebook, right before creating the `recommendation engine` itself. 

### **Popularity**

In [38]:
movies[['popularity']]

popularity
original_title                                      
Avatar                                    150.437577
Pirates of the Caribbean: At World's End  139.082615
Spectre                                   107.376788
The Dark Knight Rises                     112.312950
John Carter                                43.926995
...                                              ...
Primer                                     23.307949
Cavite                                      0.022173
Newlyweds                                   0.642552
Signed, Sealed, Delivered                   1.444476
My Date with Drew                           1.929883

[4476 rows x 1 columns]

I don't like data columns like these. How is this measured? On what scale? And according to whom? It's a set of data that's a bit too subjective for my liking.

That being said, let's drop this column from our dataset.

In [39]:
movies.drop(columns = ['popularity'], inplace = True)

### **Production Companies**

In [40]:
movies[['production_companies']]

production_companies
original_title                                                                             
Avatar                                    [{"name": "Ingenious Film Partners", "id": 289...
Pirates of the Caribbean: At World's End  [{"name": "Walt Disney Pictures", "id": 2}, {"...
Spectre                                   [{"name": "Columbia Pictures", "id": 5}, {"nam...
The Dark Knight Rises                     [{"name": "Legendary Pictures", "id": 923}, {"...
John Carter                                     [{"name": "Walt Disney Pictures", "id": 2}]
...                                                                                     ...
Primer                                                   [{"name": "Thinkfilm", "id": 446}]
Cavite                                                                                   []
Newlyweds                                                                                []
Signed, Sealed, Delivered                 [{"name": "Front Street Pictures", "id": 3958}...
My Date with Drew                         [{"name": "rusty bear entertainment", "id": 87...

[4476 rows x 1 columns]

In [41]:
# Let's look at an example of a single "production_company" cell value
movies['production_companies'][0]

'[{"name": "Ingenious Film Partners", "id": 289}, {"name": "Twentieth Century Fox Film Corporation", "id": 306}, {"name": "Dune Entertainment", "id": 444}, {"name": "Lightstorm Entertainment", "id": 574}]'

It looks like this is a lit of dictionaries that has been converted into a string. In the cell below, I'll see if it'll be possible to isolate the names of the production companies, which will be crucial to doing some feature engineering in the next notebook. 

In [42]:
# Let's used json to convert this string back to a list of dictionaries
import json

# Use the .loads methodology to 
production_companies_0 = json.loads(movies['production_companies'][0])
production_companies_0

[{'name': 'Ingenious Film Partners', 'id': 289},
 {'name': 'Twentieth Century Fox Film Corporation', 'id': 306},
 {'name': 'Dune Entertainment', 'id': 444},
 {'name': 'Lightstorm Entertainment', 'id': 574}]

In [43]:
# Create an empty list called 'production_companies'
production_companies = []

# Create a list of indices as long as the column 'production_companies' in the dataframe
for i in range(0, len(movies['production_companies'])):
    
    # Isolate each cell one by one and create a list of production_companies in the cell
    cell = movies.iloc[i]['production_companies'].split(' ')
    
    # Loop through each list to look for unique production_companies
    for production_company in cell:
        
        # Add unique production_companies to the production_companies list
        if production_company not in production_companies:
            production_companies.append(production_company)
            
# Take a look at the production_companies list and how many unique production_companies are in it
print(f'There are {len(production_companies)} unique production_companies in the dataset, and they are:')
print(production_companies)

There are 10433 unique production_companies in the dataset, and they are:
['[{"name":', '"Ingenious', 'Film', 'Partners",', '"id":', '289},', '{"name":', '"Twentieth', 'Century', 'Fox', 'Corporation",', '306},', '"Dune', 'Entertainment",', '444},', '"Lightstorm', '574}]', '"Walt', 'Disney', 'Pictures",', '2},', '"Jerry', 'Bruckheimer', 'Films",', '130},', '"Second', 'Mate', 'Productions",', '19936}]', '"Columbia', '5},', '"Danjaq",', '10761},', '"B24",', '69434}]', '"Legendary', '923},', '"Warner', 'Bros.",', '6194},', '"DC', '9993},', '"Syncopy",', '9996}]', '2}]', '"Laura', 'Ziskin', '326},', '"Marvel', 'Enterprises",', '19551}]', 'Animation', 'Studios",', '6125}]', '420},', '"Prime', 'Focus",', '15357},', '"Revolution', 'Sun', '76043}]', '"Heyday', '7364}]', 'Comics",', '429},', '"Atlas', '507},', '"Cruel', '&', 'Unusual', '9995},', '"RatPac-Dune', '41624}]', '"Bad', 'Hat', 'Harry', '9168}]', '"Eon', '7576}]', '"Infinitum', 'Nihil",', '2691},', '"Silver', 'Bullet', 'Productions', '(

This was worth exploring, but this looks like a column that should be dropped. There are `10,433` unique production companies (upon further review, it's less than that -- for instance, "New Line Cinema" was split up into "New", "Line", and "Cinema". Fixing this simply isn't worth the effort -- including these companies in the dataframe may have been valuable, but its value doesn't justify the time it'll take to work on this.

In [44]:
# Let's drop this column from the dataframe
movies.drop(columns = 'production_companies', inplace = True)

### **Production Countries**

In [45]:
movies[['production_countries']]

production_countries
original_title                                                                             
Avatar                                    [{"iso_3166_1": "US", "name": "United States o...
Pirates of the Caribbean: At World's End  [{"iso_3166_1": "US", "name": "United States o...
Spectre                                   [{"iso_3166_1": "GB", "name": "United Kingdom"...
The Dark Knight Rises                     [{"iso_3166_1": "US", "name": "United States o...
John Carter                               [{"iso_3166_1": "US", "name": "United States o...
...                                                                                     ...
Primer                                    [{"iso_3166_1": "US", "name": "United States o...
Cavite                                                                                   []
Newlyweds                                                                                []
Signed, Sealed, Delivered                 [{"iso_3166_1": "US", "name": "United States o...
My Date with Drew                         [{"iso_3166_1": "US", "name": "United States o...

[4476 rows x 1 columns]

The same is true in this column as was in the `production_companies` column. We can convert these to a `dictionary` from a string, but would it even be worth it? `production_countries` probably isn't a great predictor for a recommendation engine.

In [46]:
movies.drop(columns = 'production_countries', inplace = True)

### **Release Date**

In [47]:
movies['release_date']

original_title
Avatar                                      2009-12-10
Pirates of the Caribbean: At World's End    2007-05-19
Spectre                                     2015-10-26
The Dark Knight Rises                       2012-07-16
John Carter                                 2012-03-07
                                               ...    
Primer                                      2004-10-08
Cavite                                      2005-03-12
Newlyweds                                   2011-12-26
Signed, Sealed, Delivered                   2013-10-13
My Date with Drew                           2005-08-05
Name: release_date, Length: 4476, dtype: object

The `data type` here is an `object` and should instead be converted to a `datetime` format.

In [48]:
# Import datetime
from datetime import datetime

# Change datatype of 'release_date' column to datetime
movies['release_date'] = pd.to_datetime(movies['release_date'])

# Check work
movies['release_date']

original_title
Avatar                                     2009-12-10
Pirates of the Caribbean: At World's End   2007-05-19
Spectre                                    2015-10-26
The Dark Knight Rises                      2012-07-16
John Carter                                2012-03-07
                                              ...    
Primer                                     2004-10-08
Cavite                                     2005-03-12
Newlyweds                                  2011-12-26
Signed, Sealed, Delivered                  2013-10-13
My Date with Drew                          2005-08-05
Name: release_date, Length: 4476, dtype: datetime64[ns]

### Runtime

In [49]:
# Check format and data type
movies['runtime']

original_title
Avatar                                      162.0
Pirates of the Caribbean: At World's End    169.0
Spectre                                     148.0
The Dark Knight Rises                       165.0
John Carter                                 132.0
                                            ...  
Primer                                       77.0
Cavite                                       80.0
Newlyweds                                    85.0
Signed, Sealed, Delivered                   120.0
My Date with Drew                            90.0
Name: runtime, Length: 4476, dtype: float64

In [50]:
# There were 2 NaNs initially in this dataframe -- how many are there now?
movies['runtime'].isna().sum()

0

### Spoken Languages

In [51]:
# Check format and data type
movies['spoken_languages']

original_title
Avatar                                      [{"iso_639_1": "en", "name": "English"}, {"iso...
Pirates of the Caribbean: At World's End             [{"iso_639_1": "en", "name": "English"}]
Spectre                                     [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...
The Dark Knight Rises                                [{"iso_639_1": "en", "name": "English"}]
John Carter                                          [{"iso_639_1": "en", "name": "English"}]
                                                                  ...                        
Primer                                               [{"iso_639_1": "en", "name": "English"}]
Cavite                                                                                     []
Newlyweds                                                                                  []
Signed, Sealed, Delivered                            [{"iso_639_1": "en", "name": "English"}]
My Date with Drew                            

In [52]:
# Isolate one cell to see what its contents look like
movies['spoken_languages'][0]

'[{"iso_639_1": "en", "name": "English"}, {"iso_639_1": "es", "name": "Espa\\u00f1ol"}]'

The contents of this column are unclear. I've seen Avatar, the movie that the `spoken_languages` above corresponds to, and there's no Spanish in the movie. Could it mean that there exists a Spanish translated version of Avatar? 

Whatever the purpose of this column is, it doesn't appear to be necessary. Let's drop it.

In [53]:
movies.drop(columns = ['spoken_languages'], inplace = True)

### Status

In [54]:
# Check format and data type
movies['status']

original_title
Avatar                                      Released
Pirates of the Caribbean: At World's End    Released
Spectre                                     Released
The Dark Knight Rises                       Released
John Carter                                 Released
                                              ...   
Primer                                      Released
Cavite                                      Released
Newlyweds                                   Released
Signed, Sealed, Delivered                   Released
My Date with Drew                           Released
Name: status, Length: 4476, dtype: object

In [55]:
# Look at unique values in this column
movies['status'].unique()

array(['Released', 'Post Production', 'Rumored'], dtype=object)

`Rumored` and movies in `Post Production` are included in this? Let's take a closer look.

In [56]:
# Take a look at all movies with the status 'Post Production'
movies[movies['status'] == 'Post Production']

budget release_date  revenue  runtime           status  \
original_title                                                            
Brotherly Love  1900000   2015-04-24        0     89.0  Post Production   
Higher Ground   2000000   2011-08-26   841733    109.0  Post Production   

               tagline           title  vote_average  vote_count  \
original_title                                                     
Brotherly Love     NaN  Brotherly Love           6.9          21   
Higher Ground      NaN   Higher Ground           5.3          14   

                                                             cast  ...  \
original_title                                                     ...   
Brotherly Love  Keke Palmer Cory Hardrict Eric D. Hill Jr. Qui...  ...   
Higher Ground   Vera Farmiga Donna Murphy Dagmara Domi\u0144cz...  ...   

               Romance Horror  Mystery  History  War  Music  Documentary  \
original_title                                                             
Brotherly Love       0      0        0        0    0      0            0   
Higher Ground        0      0        0        0    0      0            0   

                Foreign  TV                                           overview  
original_title                                                                  
Brotherly Love        0   0  [west, philadelphia, basketball, star, sergio,...  
Higher Ground         0   0  [a, chronicle, of, one, woman, s, lifelong, st...  

[2 rows x 33 columns]

In [57]:
# Take a look at all movies with the status 'Rumored'
movies[movies['status'] == 'Rumored']

budget release_date  revenue  runtime   status  \
original_title                                                             
The Helix... Loaded            0   2005-01-01        0     97.0  Rumored   
Crying with Laughter           0   2009-06-01        0     93.0  Rumored   
The Harvest (La Cosecha)   56000   2011-07-29        0     80.0  Rumored   
Little Big Top                 0   2006-01-01        0      0.0  Rumored   
The Naked Ape                  0   2006-09-16        0    110.0  Rumored   

                                              tagline  \
original_title                                          
The Helix... Loaded                               NaN   
Crying with Laughter      A Bad Trip Down Memory Lane   
The Harvest (La Cosecha)                          NaN   
Little Big Top                                    NaN   
The Naked Ape                                     NaN   

                                             title  vote_average  vote_count  \
original_title                                                                 
The Helix... Loaded            The Helix... Loaded           4.8           2   
Crying with Laughter          Crying with Laughter           7.0           1   
The Harvest (La Cosecha)  The Harvest (La Cosecha)           0.0           0   
Little Big Top                      Little Big Top          10.0           1   
The Naked Ape                        The Naked Ape           5.0           1   

                                                                       cast  \
original_title                                                                
The Helix... Loaded                                                     NaN   
Crying with Laughter      Stephen McCole Malcolm Shields Andrew Neil Jo ...   
The Harvest (La Cosecha)                                                NaN   
Little Big Top            Sid Haig Richard Riehle Mel England Jacob Zach...   
The Naked Ape             Josh Wise Chelse Swain Sean Shanks Amanda MacD...   

                          ... Romance Horror  Mystery  History  War  Music  \
original_title            ...                                                
The Helix... Loaded       ...       0      0        0        0    0      0   
Crying with Laughter      ...       0      0        0        0    0      0   
The Harvest (La Cosecha)  ...       0      0        0        0    0      0   
Little Big Top            ...       0      0        0        0    0      0   
The Naked Ape             ...       0      0        0        0    0      0   

                          Documentary  Foreign  TV  \
original_title                                       
The Helix... Loaded                 0        0   0   
Crying with Laughter                0        0   0   
The Harvest (La Cosecha)            1        0   0   
Little Big Top                      0        0   0   
The Naked Ape                       0        0   0   

                                                                   overview  
original_title                                                               
The Helix... Loaded                                                      []  
Crying with Laughter      [powerfully, redemptive, and, darkly, comedic,...  
The Harvest (La Cosecha)  [the, story, of, the, child, who, work, 12, 14...  
Little Big Top            [an, aging, out, of, work, clown, return, to, ...  
The Naked Ape             [the, naked, ape, is, a, coming, of, age, film...  

[5 rows x 33 columns]

Movies with statuses that are anything except for `Released` should be droppped. Not only is a movie reccomendation for a movie that hasn't been released a crappy recommendation, but these movies look like they were never released. For example, The Naked Ape (ahem) has apparently been rumored since 2006, and a simple Google search has confirmed that there is no such film.

In [58]:
# Drop the 36 movies from the dataframe whose statuses are either 'Rumored' or 'Post Production'
movies = movies[movies['status'] == 'Released']

In [109]:
# Now let's drop the 'status' column from the dataset, now that all the data within it is uniform
movies.drop(columns = ['status'], inplace = True)

### Tagline

In [59]:
movies['tagline']

original_title
Avatar                                                            Enter the World of Pandora.
Pirates of the Caribbean: At World's End       At the end of the world, the adventure begins.
Spectre                                                                 A Plan No One Escapes
The Dark Knight Rises                                                         The Legend Ends
John Carter                                              Lost in our world, found in another.
                                                                  ...                        
Primer                                                     What happens if it actually works?
Cavite                                                                                    NaN
Newlyweds                                   A newlywed couple's honeymoon is upended by th...
Signed, Sealed, Delivered                                                                 NaN
My Date with Drew                            

This column essentially looks like a less informative and far cornier edition of the `overview` column. Let's drop it.

In [60]:
movies.drop(columns = ['tagline'], inplace = True)

That's odd that the above command generated a warning...let's check to make sure that this column is gone from `movies`.

In [61]:
movies.columns

Index(['budget', 'release_date', 'revenue', 'runtime', 'status', 'title',
       'vote_average', 'vote_count', 'cast', 'crew', 'director', 'Action',
       'Adventure', 'Fantasy', 'Science_Fiction', 'Crime', 'Drama', 'Thriller',
       'Animation', 'Family', 'Western', 'Comedy', 'Romance', 'Horror',
       'Mystery', 'History', 'War', 'Music', 'Documentary', 'Foreign', 'TV',
       'overview'],
      dtype='object')

Good, it looks like the column is gone.

### Title

How is this different from `original_title` ?

In [62]:
# Check column and data type
movies['title']

original_title
Avatar                                                                        Avatar
Pirates of the Caribbean: At World's End    Pirates of the Caribbean: At World's End
Spectre                                                                      Spectre
The Dark Knight Rises                                          The Dark Knight Rises
John Carter                                                              John Carter
                                                              ...                   
Primer                                                                        Primer
Cavite                                                                        Cavite
Newlyweds                                                                  Newlyweds
Signed, Sealed, Delivered                                  Signed, Sealed, Delivered
My Date with Drew                                                  My Date with Drew
Name: title, Length: 4469, dtype: object

Let's check to see how exactly, if it all, this column is different from `orginial_title`.

In [63]:
# Define a function to look for differences in 'original_title' and 'title'
def check_title_differences(df):
    
    # Create an empty dataframe
    title_differences = pd.DataFrame()
    
    # Create an empty list called 'original titles'
    original_titles = []
    
    # Create an empty list called 'titles'
    titles = []
    
    # Start for loop to go through the entire dataframe
    for i in range(0, len(df.index)):
        
        # If statement: if 'original title' and 'title' are unequal values
        if df.index[i] != df.iloc[i]['title']:
            
            # Add unequal original title to appropriate list
            original_titles.append(df.index[i])
            
            # Add unequal title to appropriate list
            titles.append(df.iloc[i]['title'])
    
    # Create 'original titles' column in new dataframe with newly created list
    title_differences['original_titles'] = original_titles
    
    # Create 'titles' column in new dataframe with newly created list
    title_differences['titles'] = titles
    
    # Return new dataframe
    return title_differences

In [64]:
# Call function on 'movies' dataframe
check_title_differences(movies)

original_titles  \
0             4: Rise of the Silver Surfer   
1                   Arthur et les Minimoys   
2                              Deux frères   
3             Michael Jackson's This Is It   
4                             Lo imposible   
5                                    Nomad   
6                   Le Hussard sur le toit   
7                       The House of Magic   
8                    The Neverending Story   
9                                EverAfter   
10            Highlander III: The Sorcerer   
11                     La migliore offerta   
12  Sammy's avonturen: De geheime doorgang   
13                                   Chéri   
14                                 三国之见龙卸甲   
15                                  Saw 3D   
16           Jackass Presents: Bad Grandpa   
17                  The Hills Have Eyes II   
18                                 Bathory   
19                                Autómata   
20                La veuve de Saint-Pierre   
21                      How to Be a Player   
22                          Run Fatboy Run   
23          La femme de chambre du Titanic   
24                          Hai Marjawaan!   
25                     Due occhi diabolici   
26        Star Wars: Clone Wars (Volume 1)   
27                          Håkon Håkonsen   
28  And When Did You Last See Your Father?   
29                             Hodejegerne   
30               Salvando al Soldado Perez   
31                             Dawn Patrol   
32                            Opposite Sex   
33                       Rejsen til Saturn   
34                               Mad Max 2   
35                              Jerusalema   
36                         $upercapitalist   
37                   Walking with the Dead   
38                   La Navaja De Don Juan   
39                            Cama adentro   
40           L.I.E. Long Island Expressway   
41                            Arnolds Park   
42                   Kita svajonių komanda   
43                           Chats perchés   

                                        titles  
0       Fantastic 4: Rise of the Silver Surfer  
1                    Arthur and the Invisibles  
2                                 Two Brothers  
3                                   This Is It  
4                               The Impossible  
5                           Nomad: The Warrior  
6                     The Horseman on the Roof  
7               Thunder and the House of Magic  
8                        The NeverEnding Story  
9               Ever After: A Cinderella Story  
10             Highlander: The Final Dimension  
11                              The Best Offer  
12         A Turtle's Tale: Sammy's Adventures  
13                                       Cheri  
14  Three Kingdoms: Resurrection of the Dragon  
15                      Saw: The Final Chapter  
16                                 Bad Grandpa  
17                       The Hills Have Eyes 2  
18                  Bathory: Countess of Blood  
19                                    Automata  
20                   The Widow of Saint-Pierre  
21                Def Jam's How to Be a Player  
22                            Run, Fatboy, Run  
23              The Chambermaid on the Titanic  
24                  It's a Wonderful Afterlife  
25                               Two Evil Eyes  
26             Star Wars: Clone Wars: Volume 1  
27                                 Shipwrecked  
28          When Did You Last See Your Father?  
29                                 Headhunters  
30                        Saving Private Perez  
31                                    Stranded  
32                            The Opposite Sex  
33                           Journey to Saturn  
34                 Mad Max 2: The Road Warrior  
35             Gangster's Paradise: Jerusalema  
36                             Supercapitalist  
37                        The Walking Deceased  
38                       The Blade of Don Jua

It might be a good idea to reset the index of `movies` to `titles`, simply because the column `title` seems to be entirely in English, whereas the `original_title` column (which is now the index) does not.

In [65]:
# Check columns
movies.columns

Index(['budget', 'release_date', 'revenue', 'runtime', 'status', 'title',
       'vote_average', 'vote_count', 'cast', 'crew', 'director', 'Action',
       'Adventure', 'Fantasy', 'Science_Fiction', 'Crime', 'Drama', 'Thriller',
       'Animation', 'Family', 'Western', 'Comedy', 'Romance', 'Horror',
       'Mystery', 'History', 'War', 'Music', 'Documentary', 'Foreign', 'TV',
       'overview'],
      dtype='object')

In [66]:
# Drop 'title' from the dataframe now that it's been set as the index
movies.drop(columns = 'title', inplace = True)

### Vote Average and Vote Count

In [67]:
# Preview column
movies['vote_average']

original_title
Avatar                                      7.2
Pirates of the Caribbean: At World's End    6.9
Spectre                                     6.3
The Dark Knight Rises                       7.6
John Carter                                 6.1
                                           ... 
Primer                                      6.9
Cavite                                      7.5
Newlyweds                                   5.9
Signed, Sealed, Delivered                   7.0
My Date with Drew                           6.3
Name: vote_average, Length: 4469, dtype: float64

In [68]:
# Preview column
movies['vote_count']

original_title
Avatar                                      11800
Pirates of the Caribbean: At World's End     4500
Spectre                                      4466
The Dark Knight Rises                        9106
John Carter                                  2124
                                            ...  
Primer                                        658
Cavite                                          2
Newlyweds                                       5
Signed, Sealed, Delivered                       6
My Date with Drew                              16
Name: vote_count, Length: 4469, dtype: int64

These are much better indicators of popularity than the obscure column titled `popularity` that was dropped previously.

### Cast

In [69]:
# Preview column
movies['cast']

original_title
Avatar                                      Sam Worthington Zoe Saldana Sigourney Weaver S...
Pirates of the Caribbean: At World's End    Johnny Depp Orlando Bloom Keira Knightley Stel...
Spectre                                     Daniel Craig Christoph Waltz L\u00e9a Seydoux ...
The Dark Knight Rises                       Christian Bale Michael Caine Gary Oldman Anne ...
John Carter                                 Taylor Kitsch Lynn Collins Samantha Morton Wil...
                                                                  ...                        
Primer                                      Shane Carruth David Sullivan Casey Gooden Anan...
Cavite                                                                                    NaN
Newlyweds                                   Edward Burns Kerry Bish\u00e9 Marsha Dietlein ...
Signed, Sealed, Delivered                   Eric Mabius Kristin Booth Crystal Lowe Geoff G...
My Date with Drew                           D

In [70]:
# Isolate a single cell in this column for a preview
movies['cast'][0]

'Sam Worthington Zoe Saldana Sigourney Weaver Stephen Lang Michelle Rodriguez'

This is a difficult column to work with. While I believe a movie's cast can be a good indicator when it comes to making movie recommendations, there are likely far too many actors in this dataset to account for. Is it possible? Yes. But it would mean that there were more columns than rows in this dataset, which is never a best practice. 

Perhaps what we can do it just isolate the first actor in the dataset and drop the rest of the cast from the column, thus renaming the column from `cast` to `star`.

Before doing so, we need to see if there are any `NaNs` in this column, or else the `.split()` method will break the for loop.

In [71]:
# Number of nulls for 'cast' column
movies['cast'].isnull().sum()

24

In [72]:
# Let's take a look at all these movies
movies[movies['cast'].isna()]

budget release_date   revenue  \
original_title                                                          
Barney's Great Adventure                     0   1998-01-01         0   
Running Forever                              0   2015-10-27         0   
2016: Obama's America                  2500000   2012-07-13  33400000   
Me You and Five Bucks                        1   2015-07-07         0   
Down & Out With The Dolls                    0   2003-03-20         0   
Sex With Strangers                           0   2002-02-22         0   
The Hadza:  Last of the First                0   2014-10-31         0   
Mad Hot Ballroom                             0   2005-05-13         0   
Straight Out of Brooklyn                     0   1991-06-13         0   
I Want Your Money                            0   2010-07-06         0   
Fabled                                       0   2002-01-01         0   
Sisters in Law                               0   2005-09-03         0   
Ayurveda: Art of Being                  300000   2001-09-20   2074000   
Amidst the Devil's Wings                300000   2014-12-01         0   
Call + Response                              0   2008-10-10         0   
Chats perchés                                0   2004-12-05         0   
The Trials Of Darryl Hunt                    0   2007-06-15         0   
Butterfly Girl                               0   2014-03-10         0   
Antarctic Edge: 70° South                    0   2015-04-17         0   
Top Spin                                     0   2014-11-15         0   
Peace, Propaganda & the Promised Land        0   2005-01-28         0   
Counting                                 50000   2015-02-09         0   
Bending Steel                                0   2013-04-20         0   
Cavite                                       0   2005-03-12         0   

                                       runtime    status  vote_average  \
original_title                                                           
Barney's Great Adventure                  76.0  Released           2.9   
Running Forever                           88.0  Released           0.0   
2016: Obama's America                     87.0  Released           4.6   
Me You and Five Bucks                     90.0  Released          10.0   
Down & Out With The Dolls                 88.0  Released           0.0   
Sex With Strangers                         0.0  Released           5.0   
The Hadza:  Last of the First             70.0  Released           0.0   
Mad Hot Ballroom                         105.0  Released           7.2   
Straight Out of Brooklyn                   0.0  Released           4.3   
I Want Your Money                         92.0  Released           3.8   
Fabled                                    84.0  Released           0.0   
Sisters in Law                           104.0  Released           7.3   
Ayurveda: Art of Being                   101.0  Released           5.5   
Amidst the Devil's Wings                  90.0  Released           0.0   
Call + Response                           86.0  Released           8.0   
Chats perchés                             59.0  Released           7.7   
The Trials Of Darryl Hunt                106.0  Released           6.8   
Butterfly Girl                            77.0  Released           0.0   
Antarctic Edge: 70° South                 72.0  Released           0.0   
Top Spin                                  80.0  Released           6.8   
Peace, Propaganda & the Promised Land     80.0  Released           6.4   
Counting                                 111.0  Released           8.3   
Bending Steel                             93.0  Released           5.0   
Cavite                                    80.0  Released           7.5   

                                       vote_count cast  \
original_title                                           
Barney's Great Adventure                        8  NaN   
Running Forever                           

QUESTION: Will anyone care if we drop these castless movies from our dataset?

In [73]:
# Tally total number of votew for all 24 castless movies
movies[movies['cast'].isna()]['vote_count'].sum()

76

ANSWER: probably not.

In [74]:
movies = movies[movies['cast'].isna() == False]

#### Cleaning the Cast Column

The cast column is a single string containing the names of the biggest names in each movie. Some of the names have errors in them, containing characters that aren't letters. In the cell below, we'll run a command to change each string into a list of names and to fix any spelling errors.

In [95]:
# Create an empty list
cleaned_names = []

# Loop through all values in the 'cast' column
for i in movies['cast']:
    
    # Create an empty list for each movie in the loop
    cast = []
    
    # Turn the string into a list of names, splitting on the spaces
    cell = i.split(' ')
    
    # Loop through each name in the variable 'cell'
    for name in cell:
        
        # Create an empty string object
        cleaned_name = ''
        
        # Add each character to 'cleaned_name' if it is, in fact, a letter
        for letter in name:
            if letter.isalpha() == True: cleaned_name += letter 
         
        # Add the cleaned name to 'cast'
        cast.append(cleaned_name)
        
    # Add the new cleaned cell to the 'cleaned_names' list
    cleaned_names.append(cast)
            
# Check work          
cleaned_names

[['Sam',
  'Worthington',
  'Zoe',
  'Saldana',
  'Sigourney',
  'Weaver',
  'Stephen',
  'Lang',
  'Michelle',
  'Rodriguez'],
 ['Johnny',
  'Depp',
  'Orlando',
  'Bloom',
  'Keira',
  'Knightley',
  'Stellan',
  'Skarsguerd',
  'Chow',
  'Yunfat'],
 ['Daniel',
  'Craig',
  'Christoph',
  'Waltz',
  'Luea',
  'Seydoux',
  'Ralph',
  'Fiennes',
  'Monica',
  'Bellucci'],
 ['Christian',
  'Bale',
  'Michael',
  'Caine',
  'Gary',
  'Oldman',
  'Anne',
  'Hathaway',
  'Tom',
  'Hardy'],
 ['Taylor',
  'Kitsch',
  'Lynn',
  'Collins',
  'Samantha',
  'Morton',
  'Willem',
  'Dafoe',
  'Thomas',
  'Haden',
  'Church'],
 ['Tobey',
  'Maguire',
  'Kirsten',
  'Dunst',
  'James',
  'Franco',
  'Thomas',
  'Haden',
  'Church',
  'Topher',
  'Grace'],
 ['Zachary',
  'Levi',
  'Mandy',
  'Moore',
  'Donna',
  'Murphy',
  'Ron',
  'Perlman',
  'MC',
  'Gainey'],
 ['Robert',
  'Downey',
  'Jr',
  'Chris',
  'Hemsworth',
  'Mark',
  'Ruffalo',
  'Chris',
  'Evans',
  'Scarlett',
  'Johansson'],
 ['

In [96]:
# Replace old cast column with new move
movies['cast'] = cleaned_names

### **Budget**

In [97]:
# Preview column budget
movies['budget']

original_title
Avatar                                      237000000
Pirates of the Caribbean: At World's End    300000000
Spectre                                     245000000
The Dark Knight Rises                       250000000
John Carter                                 260000000
                                              ...    
Bang                                                0
Primer                                           7000
Newlyweds                                        9000
Signed, Sealed, Delivered                           0
My Date with Drew                                   0
Name: budget, Length: 4445, dtype: int64

The data in this column is difficult to read due to the lack of `,` seperators. Let's assume the data is in American dollars as well.

Some of these values are for `$0` -- are these movies created by studios in a kind of pro bono fashion, or do the `$0` values represent missing data?

In [98]:
movies[movies['budget'] == 0].shape

(874, 31)

As it turns out, `1037` of the values in the `budget` column are `$0`. 

These scenarios aren't ideal. One idea is to drop this column entirely from the dataset, but I think a film's `budget` would be a good prediction factor to include in the recommendation engine. For example, some people might prefer blockbuster movies with blockbuster budgets, while other people prefer indie movies that with modest budgets.

There are several tactics for tackling this problem. For this particular column, I'm going to fill in all of the missing values with the `mean` value of all the budget amounts in the dataset that are not equal to zero.

In [99]:
"${:,.2f}".format(round(sum(movies[movies['budget'] != 0]['budget']) / len(movies[movies['budget'] != 0]['budget']), 
                 2))

'$38,341,022.04'

The average budget for a movie in this daatset is `$37,042,837.63`. We'll now use this value to replace all the `$0s` in this dataset within the `budget` column.

In [100]:
# Assign average budget to a variable
mean_budget = sum(movies[movies['budget'] != 0]['budget']) / len(movies[movies['budget'] != 0]['budget'])

# Define a function to take in a dataframe
def replace_0_budget(df):
    
    # Create an empty list that will be the new column
    new_budget_column = []
    
    # Begin for loop
    for i in df['budget']:
        
        # If budget doesn't equal 0
        if i != 0:
            
            # Append existing value to empty list
            new_budget_column.append(i)
        
        # If budget is equal to zero
        else:
            
            # Append average budget to new column instead of 0
            new_budget_column.append(mean_budget)
    
    # Replace old column with new column
    df['budget'] = new_budget_column        
    
    # Return dataframe
    return df

# Call function on the movies dataframe
replace_0_budget(movies)

budget release_date  \
original_title                                                        
Avatar                                    2.370000e+08   2009-12-10   
Pirates of the Caribbean: At World's End  3.000000e+08   2007-05-19   
Spectre                                   2.450000e+08   2015-10-26   
The Dark Knight Rises                     2.500000e+08   2012-07-16   
John Carter                               2.600000e+08   2012-03-07   
...                                                ...          ...   
Bang                                      3.834102e+07   1995-09-09   
Primer                                    7.000000e+03   2004-10-08   
Newlyweds                                 9.000000e+03   2011-12-26   
Signed, Sealed, Delivered                 3.834102e+07   2013-10-13   
My Date with Drew                         3.834102e+07   2005-08-05   

                                             revenue  runtime    status  \
original_title                                                            
Avatar                                    2787965087    162.0  Released   
Pirates of the Caribbean: At World's End   961000000    169.0  Released   
Spectre                                    880674609    148.0  Released   
The Dark Knight Rises                     1084939099    165.0  Released   
John Carter                                284139100    132.0  Released   
...                                              ...      ...       ...   
Bang                                               0     98.0  Released   
Primer                                        424760     77.0  Released   
Newlyweds                                          0     85.0  Released   
Signed, Sealed, Delivered                          0    120.0  Released   
My Date with Drew                                  0     90.0  Released   

                                          vote_average  vote_count  \
original_title                                                       
Avatar                                             7.2       11800   
Pirates of the Caribbean: At World's End           6.9        4500   
Spectre                                            6.3        4466   
The Dark Knight Rises                              7.6        9106   
John Carter                                        6.1        2124   
...                                                ...         ...   
Bang                                               6.0           1   
Primer                                             6.9         658   
Newlyweds                                          5.9           5   
Signed, Sealed, Delivered                          7.0           6   
My Date with Drew                                  6.3          16   

                                                                                       cast  \
original_title                                                                                
Avatar                                    [Sam, Worthington, Zoe, Saldana, Sigourney, We...   
Pirates of the Caribbean: At World's End  [Johnny, Depp, Orlando, Bloom, Keira, Knightle...   
Spectre                                   [Daniel, Craig, Christoph, Waltz, Luea, Seydou...   
The Dark Knight Rises                     [Christian, Bale, Michael, Caine, Gary, Oldman...   
John Carter                               [Taylor, Kitsch, Lynn, Collins, Samantha, Mort...   
...                                                                                     ...   
Bang                                      [Darling, Narita, Peter, Greene, Michael, Newl...   
Primer                                    [Shane, Carruth, David, Sullivan, Casey, Goode...   
Newlyweds                                 [Edward, Burns, Kerry, Bishue, Marsha, Dietlei...   
Signed, Sealed, Delivered                 [Eric, Mabius, Kristin, Booth, Crystal, Lowe, ...   
My Date with Drew                         [Drew, Barrymore, Brian, Herzlinger, Corey, Fe...   

 

In [101]:
movies[movies['budget'] == 0].shape

(0, 31)

Huzzah!! All movies with a `$0` budget have now had their value in the dataset changed to the average budget of all budgets in the dataset.

### **Revenue**

In [102]:
movies[['revenue']]

revenue
original_title                                      
Avatar                                    2787965087
Pirates of the Caribbean: At World's End   961000000
Spectre                                    880674609
The Dark Knight Rises                     1084939099
John Carter                                284139100
...                                              ...
Bang                                               0
Primer                                        424760
Newlyweds                                          0
Signed, Sealed, Delivered                          0
My Date with Drew                                  0

[4445 rows x 1 columns]

The values in this column are a little hard to read without any comma seperators.

It looks like some movies have made `$0` in revenue, which is difficult to believe. How many movies in our dataframe have posted `$0` in revenue?

In [103]:
movies[movies['revenue'] == 0].shape

(1216, 31)

Hmmm, so `1,244` movies haven't made a dime? This seems highly implausible.



In [104]:
"${:,.2f}".format(round(sum(movies[movies['revenue'] != 0]['revenue']) / len(movies[movies['revenue'] != 0]['revenue']), 
                 2))

'$120,640,826.68'

The average amount of revenue earned by a movie in this daatset is `$120,540,082.02`. We'll now use this value to replace all the `$0s` in this dataset within the `revenue` column.

In [105]:
mean_revenue = sum(movies[movies['revenue'] != 0]['revenue']) / len(movies[movies['revenue'] != 0]['revenue'])

def replace_0_revenue(df):
    new_revenue_column = []
    for i in df['revenue']:
        if i != 0:
            new_revenue_column.append(i)
        else:
            new_revenue_column.append(mean_revenue)
    df['revenue'] = new_revenue_column        
    return df

replace_0_revenue(movies)

budget release_date  \
original_title                                                        
Avatar                                    2.370000e+08   2009-12-10   
Pirates of the Caribbean: At World's End  3.000000e+08   2007-05-19   
Spectre                                   2.450000e+08   2015-10-26   
The Dark Knight Rises                     2.500000e+08   2012-07-16   
John Carter                               2.600000e+08   2012-03-07   
...                                                ...          ...   
Bang                                      3.834102e+07   1995-09-09   
Primer                                    7.000000e+03   2004-10-08   
Newlyweds                                 9.000000e+03   2011-12-26   
Signed, Sealed, Delivered                 3.834102e+07   2013-10-13   
My Date with Drew                         3.834102e+07   2005-08-05   

                                               revenue  runtime    status  \
original_title                                                              
Avatar                                    2.787965e+09    162.0  Released   
Pirates of the Caribbean: At World's End  9.610000e+08    169.0  Released   
Spectre                                   8.806746e+08    148.0  Released   
The Dark Knight Rises                     1.084939e+09    165.0  Released   
John Carter                               2.841391e+08    132.0  Released   
...                                                ...      ...       ...   
Bang                                      1.206408e+08     98.0  Released   
Primer                                    4.247600e+05     77.0  Released   
Newlyweds                                 1.206408e+08     85.0  Released   
Signed, Sealed, Delivered                 1.206408e+08    120.0  Released   
My Date with Drew                         1.206408e+08     90.0  Released   

                                          vote_average  vote_count  \
original_title                                                       
Avatar                                             7.2       11800   
Pirates of the Caribbean: At World's End           6.9        4500   
Spectre                                            6.3        4466   
The Dark Knight Rises                              7.6        9106   
John Carter                                        6.1        2124   
...                                                ...         ...   
Bang                                               6.0           1   
Primer                                             6.9         658   
Newlyweds                                          5.9           5   
Signed, Sealed, Delivered                          7.0           6   
My Date with Drew                                  6.3          16   

                                                                                       cast  \
original_title                                                                                
Avatar                                    [Sam, Worthington, Zoe, Saldana, Sigourney, We...   
Pirates of the Caribbean: At World's End  [Johnny, Depp, Orlando, Bloom, Keira, Knightle...   
Spectre                                   [Daniel, Craig, Christoph, Waltz, Luea, Seydou...   
The Dark Knight Rises                     [Christian, Bale, Michael, Caine, Gary, Oldman...   
John Carter                               [Taylor, Kitsch, Lynn, Collins, Samantha, Mort...   
...                                                                                     ...   
Bang                                      [Darling, Narita, Peter, Greene, Michael, Newl...   
Primer                                    [Shane, Carruth, David, Sullivan, Casey, Goode...   
Newlyweds                                 [Edward, Burns, Kerry, Bishue, Marsha, Dietlei...   
Signed, Sealed, Delivered                 [Eric, Mabius, Kristin, Booth, Crystal, Lowe, ...   
My Date with Drew                         [Drew, Barrymore, Brian, Herz

In [106]:
movies[movies['revenue'] == 0].shape

(0, 31)

It worked! All movies with a revenue of `$0` has now been replaced with the mean revenue in the dataset of `$120,540,082.02`.

In [110]:
movies.columns

Index(['budget', 'release_date', 'revenue', 'runtime', 'vote_average',
       'vote_count', 'cast', 'crew', 'director', 'Action', 'Adventure',
       'Fantasy', 'Science_Fiction', 'Crime', 'Drama', 'Thriller', 'Animation',
       'Family', 'Western', 'Comedy', 'Romance', 'Horror', 'Mystery',
       'History', 'War', 'Music', 'Documentary', 'Foreign', 'TV', 'overview'],
      dtype='object')

#### Director

In [111]:
# Preview director column
movies['director'].isna().sum()

original_title
Avatar                                          James Cameron
Pirates of the Caribbean: At World's End       Gore Verbinski
Spectre                                            Sam Mendes
The Dark Knight Rises                       Christopher Nolan
John Carter                                    Andrew Stanton
                                                  ...        
Bang                                          Ash Baron-Cohen
Primer                                          Shane Carruth
Newlyweds                                        Edward Burns
Signed, Sealed, Delivered                         Scott Smith
My Date with Drew                            Brian Herzlinger
Name: director, Length: 4445, dtype: object

In [112]:
# How many movies are there with null values in the 'director' column?
movies['director'].isna().sum()

7

In [115]:
# Drop nulls from movie dataset
movies = movies[movies['director'].isna() == False]

#### Crew

In [116]:
movies['crew']

original_title
Avatar                                      [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...
Pirates of the Caribbean: At World's End    [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...
Spectre                                     [{'name': 'Thomas Newman', 'gender': 2, 'depar...
The Dark Knight Rises                       [{'name': 'Hans Zimmer', 'gender': 2, 'departm...
John Carter                                 [{'name': 'Andrew Stanton', 'gender': 2, 'depa...
                                                                  ...                        
Bang                                        [{'name': 'Ash Baron-Cohen', 'gender': 0, 'dep...
Primer                                      [{'name': 'Shane Carruth', 'gender': 2, 'depar...
Newlyweds                                   [{'name': 'Edward Burns', 'gender': 2, 'depart...
Signed, Sealed, Delivered                   [{'name': 'Carla Hetland', 'gender': 0, 'depar...
My Date with Drew                           [

In [117]:
movies.drop(columns = ['crew'], inplace = True)

/Users/michaellaccavole/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [118]:
movies

budget release_date  \
original_title                                                        
Avatar                                    2.370000e+08   2009-12-10   
Pirates of the Caribbean: At World's End  3.000000e+08   2007-05-19   
Spectre                                   2.450000e+08   2015-10-26   
The Dark Knight Rises                     2.500000e+08   2012-07-16   
John Carter                               2.600000e+08   2012-03-07   
...                                                ...          ...   
Bang                                      3.834102e+07   1995-09-09   
Primer                                    7.000000e+03   2004-10-08   
Newlyweds                                 9.000000e+03   2011-12-26   
Signed, Sealed, Delivered                 3.834102e+07   2013-10-13   
My Date with Drew                         3.834102e+07   2005-08-05   

                                               revenue  runtime  vote_average  \
original_title                                                                  
Avatar                                    2.787965e+09    162.0           7.2   
Pirates of the Caribbean: At World's End  9.610000e+08    169.0           6.9   
Spectre                                   8.806746e+08    148.0           6.3   
The Dark Knight Rises                     1.084939e+09    165.0           7.6   
John Carter                               2.841391e+08    132.0           6.1   
...                                                ...      ...           ...   
Bang                                      1.206408e+08     98.0           6.0   
Primer                                    4.247600e+05     77.0           6.9   
Newlyweds                                 1.206408e+08     85.0           5.9   
Signed, Sealed, Delivered                 1.206408e+08    120.0           7.0   
My Date with Drew                         1.206408e+08     90.0           6.3   

                                          vote_count  \
original_title                                         
Avatar                                         11800   
Pirates of the Caribbean: At World's End        4500   
Spectre                                         4466   
The Dark Knight Rises                           9106   
John Carter                                     2124   
...                                              ...   
Bang                                               1   
Primer                                           658   
Newlyweds                                          5   
Signed, Sealed, Delivered                          6   
My Date with Drew                                 16   

                                                                                       cast  \
original_title                                                                                
Avatar                                    [Sam, Worthington, Zoe, Saldana, Sigourney, We...   
Pirates of the Caribbean: At World's End  [Johnny, Depp, Orlando, Bloom, Keira, Knightle...   
Spectre                                   [Daniel, Craig, Christoph, Waltz, Luea, Seydou...   
The Dark Knight Rises                     [Christian, Bale, Michael, Caine, Gary, Oldman...   
John Carter                               [Taylor, Kitsch, Lynn, Collins, Samantha, Mort...   
...                                                                                     ...   
Bang                                      [Darling, Narita, Peter, Greene, Michael, Newl...   
Primer                                    [Shane, Carruth, David, Sullivan, Casey, Goode...   
Newlyweds                                 [Edward, Burns, Kerry, Bishue, Marsha, Dietlei...   
Signed, Sealed, Delivered                 [Eric, Mabius, Kristin, Booth, Crystal, Lowe, ...   
My Date with Drew                         [Drew, Barrymore, Brian, Herzlinger, Corey, Fe...   

                                                   director  Action  \
original_title                    

In [ ]:
pd.to_csv